packages/sdk/client-services/src/packlets/invitations/invitation-guest-extenstion.ts

```diff
@@ -0,0 +1,129 @@
+import { cancelWithContext, Context } from '@dxos/context';
+import { invariant } from '@dxos/invariant';
+import { log } from '@dxos/log';
+import { InvalidInvitationExtensionRoleError, schema } from '@dxos/protocols';
+import { type Invitation } from '@dxos/protocols/proto/dxos/client/services';
+import { type InvitationHostService, Options } from '@dxos/protocols/proto/dxos/halo/invitations';
+import { type ExtensionContext, RpcExtension } from '@dxos/teleport';
+
+const OPTIONS_TIMEOUT = 10_000;
+
+type InvitationGuestExtensionCallbacks = {
+  // Deliberately not async to not block the extensions opening.
+  onOpen: (ctx: Context, extensionCtx: ExtensionContext) => void;
+  onError: (error: Error) => void;
+
+  onStateUpdate: (newState: Invitation.State) => void;
+};
+
+/**
+ * Guest's side for a connection to a concrete peer in p2p network during invitation.
+ */
+export class InvitationGuestExtension extends RpcExtension<
+  { InvitationHostService: InvitationHostService },
+  { InvitationHostService: InvitationHostService }
+> {
+  private _ctx = new Context();
+  private _remoteOptions?: Options;
+  private _remoteOptionsTrigger = new Trigger();
+  /**
+   * Held to allow only one invitation flow at a time to be active.
+   */
+  private _invitationFlowLock: MutexGuard | null = null;
+
+  constructor(
+    private readonly _invitationFlowMutex: Mutex,
+    private readonly _callbacks: InvitationGuestExtensionCallbacks,
+  ) {
+    super({
+      requested: {
+        InvitationHostService: schema.getService('dxos.halo.invitations.InvitationHostService'),
+      },
+      exposed: {
+        InvitationHostService: schema.getService('dxos.halo.invitations.InvitationHostService'),
+      },
+    });
+  }
+
+  public hasFlowLock(): boolean {
+    return this._invitationFlowLock != null;
+  }
+
+  protected override async getHandlers(): Promise<{ InvitationHostService: InvitationHostService }> {
+    return {
+      InvitationHostService: {
+        options: async (options) => {
+          invariant(!this._remoteOptions, 'Remote options already set.');
+          this._remoteOptions = options;
+          this._remoteOptionsTrigger.wake();
+        },
+        introduce: () => {
+          throw new Error('Method not allowed.');
+        },
+        authenticate: () => {
+          throw new Error('Method not allowed.');
+        },
+        admit: () => {
+          throw new Error('Method not allowed.');
+        },
+      },
+    };
+  }
+
+  override async onOpen(context: ExtensionContext) {
+    await super.onOpen(context);
+
+    try {
+      log('guest acquire lock');
+      this._invitationFlowLock = await acquireInContext(this._ctx, this._invitationFlowMutex, {
+        releaseOnDispose: false,
+      });
+      log('guest lock acquired');
+      await cancelWithContext(this._ctx, this.rpc.InvitationHostService.options({ role: Options.Role.GUEST }));
```

Comment:
I think we should because otherwise guest won't know we're in a flow with another guest and won't service them, so they'll similarly stall, just on a later stage.
the problem is how the code is structured: part of the flow is in extension, but part is in handler. I also agree about xstate, but I didn't want to start a major refactoring cause these changes are quite big and complicated already and the vision for halo2.0 isn't clear yet (local-first-auth?), so this code might soon change again